### PR TITLE
Migrate user functions folder

### PR DESCRIPTION
**Pull request by petersg83**

### CODE REVIEW

File: examples/getstarted/src/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+module.exports = {
+  /**
+   * An asynchronous register function that runs before
+   * your application is loaded.
```
Comment (alexandrebodin) [inline examples/getstarted/src/index.js:6]: All the files are already loaded but the internal service haven't been initialized yet.

```suggestion
   * your application is initialized.
```

File: packages/core/strapi/lib/core/registries/cron.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+const cron = require('node-schedule');
+const { isFunction, getOr } = require('lodash/fp');
+
+const cronRegistry = () => {
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/cron.js:6]: Just some semantic, but I don't think this should be a registry but a service in the ./services folder instead. 
The difference is that when you add a cron you actually run when in the registries you only "store" the things.
I would have created a cronRunner like the webhooksRunner instead.

File: packages/core/strapi/lib/Strapi.js [hunk 3]
```diff
@@ -418,21 +425,14 @@ class Strapi {
...
-    const configPath = `functions.${lifecycleName}`;
-
     // plugins
     await this.container.get('modules')[lifecycleName]();
 
     // user
-    await execLifecycle(this.config.get(configPath));
+    const lifecycleFunction = this.srcIndex[lifecycleName];
+    if (lifecycleFunction) {
+      lifecycleFunction({ strapi: this });
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:434]: awai
```suggestion
      await lifecycleFunction({ strapi: this });
```

File: packages/core/strapi/lib/Strapi.js [hunk 4]
```diff
@@ -367,6 +371,9 @@ class Strapi {
 
     await this.runLifecyclesFunctions(LIFECYCLES.BOOTSTRAP);
 
+    const cronTasks = this.config.get('server.cron.tasks');
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:374]: I would actually register the crons before bootstrap and then do a cronRunner.start()
So you can add crons programmatically in the register. without them starting. and then your start everything.

File: packages/core/strapi/lib/core/registries/cron.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          {
+            rule: taskExpression,
+            ...options,
+          },
+          taskValue.task(strapi)
+        );
+        jobs.push(job);
+      }
+    },
+    cancelAll() {
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/cron.js:35]: can we make that generic and call it destroy ? as an abstraction we just want to destroy the cron service whatever it does to clean itself up is not relevant outside.

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/core/registries/cron.js (hunk 2 above)
Comment (petersg83) [reply packages/core/strapi/lib/core/registries/cron.js:6]: Totally agree

File: examples/getstarted/config/server.js [hunk 6]
```diff
@@ -7,4 +11,8 @@ module.exports = ({ env }) => ({
       secret: env('ADMIN_JWT_SECRET', 'example-token'),
     },
   },
+  cron: {
+    enabled: true,
```
Comment (alexandrebodin) [inline examples/getstarted/config/server.js:15]: I don't think we need the enabled anymore ?

File: packages/core/strapi/lib/Strapi.js [hunk 7]
```diff
@@ -330,6 +333,10 @@ class Strapi {
       entityValidator: this.entityValidator,
     });
 
+    this.cron = createCronService();
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:336]: can we create the service in the constructor instead ?

File: packages/core/strapi/lib/Strapi.js [hunk 8]
```diff
@@ -418,21 +427,14 @@ class Strapi {
...
-    };
-
-    const configPath = `functions.${lifecycleName}`;
-
     // plugins
     await this.container.get('modules')[lifecycleName]();
 
     // user
-    await execLifecycle(this.config.get(configPath));
+    const lifecycleFunction = this.srcIndex[lifecycleName];
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:434]: we could load them in this.lifecycles or sth like this.app  ?

File: packages/core/strapi/lib/services/cron.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      groupsOfTasks.push(newTasks);
+    },
+    start() {
+      for (const tasks of groupsOfTasks) {
+        for (const taskExpression in tasks) {
+          const taskValue = tasks[taskExpression];
+
+          if (isFunction(taskValue)) {
+            const instanciatedTask = taskValue(strapi);
+            return cron.scheduleJob(taskExpression, instanciatedTask);
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/cron.js:25]: need to push the job here too

File: packages/core/strapi/lib/services/cron.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+      groupsOfTasks.push(newTasks);
+    },
+    start() {
+      for (const tasks of groupsOfTasks) {
+        for (const taskExpression in tasks) {
+          const taskValue = tasks[taskExpression];
+
+          if (isFunction(taskValue)) {
+            const instanciatedTask = taskValue(strapi);
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/cron.js:24]: I think the function should be passed directly there are no  instantiation phase ?

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/services/cron.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      }
+      jobsSpecs.forEach(({ job, options }) => job.schedule(options));
+      return this;
+    },
+    stop() {
+      jobsSpecs.forEach(({ job }) => job.cancel());
+      return this;
+    },
+    destroy() {
+      return this.stop();
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/cron.js:47]: jobSpecs = [] too to avoid memory leaks

**Review by alexandrebodin — APPROVED** [review]
LGTM. Cool to see that clenaup

**Pull request merged**